# Machine Learning Engineer Nanodegree
## Supervised Learning
## Project: Building a Student Intervention System

Welcome to the second project of the Machine Learning Engineer Nanodegree! In this notebook, some template code has already been provided for you, and it will be your job to implement the additional functionality necessary to successfully complete this project. Sections that begin with **'Implementation'** in the header indicate that the following block of code will require additional functionality which you must provide. Instructions will be provided for each section and the specifics of the implementation are marked in the code block with a `'TODO'` statement. Please be sure to read the instructions carefully!

In addition to implementing code, there will be questions that you must answer which relate to the project and your implementation. Each section where you will answer a question is preceded by a **'Question X'** header. Carefully read each question and provide thorough answers in the following text boxes that begin with **'Answer:'**. Your project submission will be evaluated based on your answers to each of the questions and the implementation you provide.  

>**Note:** Code and Markdown cells can be executed using the **Shift + Enter** keyboard shortcut. In addition, Markdown cells can be edited by typically double-clicking the cell to enter edit mode.

### Question 1 - Classification vs. Regression
*Your goal for this project is to identify students who might need early intervention before they fail to graduate. Which type of supervised learning problem is this, classification or regression? Why?*

**Answer: **
This is a classification problem because there are only two classes of students that we need to be able to distinguish between: those who will graduate successfully on their own and those who need intervention to graduate. Regression problem typically deals with predicting continuous numeric targets, not labels of classes.

## Exploring the Data
Run the code cell below to load necessary Python libraries and load the student data. Note that the last column from this dataset, `'passed'`, will be our target label (whether the student graduated or didn't graduate). All other columns are features about each student.

In [1]:
# Import libraries
import numpy as np
import pandas as pd
from time import time
from sklearn.metrics import f1_score

# Read student data
student_data = pd.read_csv("student-data.csv")
print "Student data read successfully!"

Student data read successfully!


### Implementation: Data Exploration
Let's begin by investigating the dataset to determine how many students we have information on, and learn about the graduation rate among these students. In the code cell below, you will need to compute the following:
- The total number of students, `n_students`.
- The total number of features for each student, `n_features`.
- The number of those students who passed, `n_passed`.
- The number of those students who failed, `n_failed`.
- The graduation rate of the class, `grad_rate`, in percent (%).


In [2]:
#print student_data.head()
# TODO: Calculate number of students
n_students = student_data.shape[0]

# TODO: Calculate number of features
n_features = student_data.shape[1]-1

# TODO: Calculate passing students
n_passed = student_data.passed.value_counts().yes

# TODO: Calculate failing students
n_failed = student_data.passed.value_counts().no

# TODO: Calculate graduation rate
grad_rate = float(n_passed)/float(n_students)*100

# Print the results
print "Total number of students: {}".format(n_students)
print "Number of features: {}".format(n_features)
print "Number of students who passed: {}".format(n_passed)
print "Number of students who failed: {}".format(n_failed)
print "Graduation rate of the class: {:.2f}%".format(grad_rate)

Total number of students: 395
Number of features: 30
Number of students who passed: 265
Number of students who failed: 130
Graduation rate of the class: 67.09%


## Preparing the Data
In this section, we will prepare the data for modeling, training and testing.

### Identify feature and target columns
It is often the case that the data you obtain contains non-numeric features. This can be a problem, as most machine learning algorithms expect numeric data to perform computations with.

Run the code cell below to separate the student data into feature and target columns to see if any features are non-numeric.

In [3]:
# Extract feature columns
feature_cols = list(student_data.columns[:-1])

# Extract target column 'passed'
target_col = student_data.columns[-1] 

# Show the list of columns
print "Feature columns:\n{}".format(feature_cols)
print "\nTarget column: {}".format(target_col)

# Separate the data into feature data and target data (X_all and y_all, respectively)
X_all = student_data[feature_cols]
y_all = student_data[target_col]

# Show the feature information by printing the first five rows
print "\nFeature values:"
print X_all.head()

Feature columns:
['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu', 'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']

Target column: passed

Feature values:
  school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

    ...    higher internet  romantic  famrel  freetime goout Dalc Walc health  \
0   ...       yes       no        no       4         3     4    1    1      3   
1   ...       

### Preprocess Feature Columns

As you can see, there are several non-numeric columns that need to be converted! Many of them are simply `yes`/`no`, e.g. `internet`. These can be reasonably converted into `1`/`0` (binary) values.

Other columns, like `Mjob` and `Fjob`, have more than two values, and are known as _categorical variables_. The recommended way to handle such a column is to create as many columns as possible values (e.g. `Fjob_teacher`, `Fjob_other`, `Fjob_services`, etc.), and assign a `1` to one of them and `0` to all others.

These generated columns are sometimes called _dummy variables_, and we will use the [`pandas.get_dummies()`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html?highlight=get_dummies#pandas.get_dummies) function to perform this transformation. Run the code cell below to perform the preprocessing routine discussed in this section.

In [4]:
def preprocess_features(X):
    ''' Preprocesses the student data and converts non-numeric binary variables into
        binary (0/1) variables. Converts categorical variables into dummy variables. '''
    
    # Initialize new output DataFrame
    output = pd.DataFrame(index = X.index)

    # Investigate each feature column for the data
    for col, col_data in X.iteritems():
        
        # If data type is non-numeric, replace all yes/no values with 1/0
        if col_data.dtype == object:
            col_data = col_data.replace(['yes', 'no'], [1, 0])

        # If data type is categorical, convert to dummy variables
        if col_data.dtype == object:
            # Example: 'school' => 'school_GP' and 'school_MS'
            col_data = pd.get_dummies(col_data, prefix = col)  
        
        # Collect the revised columns
        output = output.join(col_data)
    
    return output

X_all = preprocess_features(X_all)
print "Processed feature columns ({} total features):\n{}".format(len(X_all.columns), list(X_all.columns))

Processed feature columns (48 total features):
['school_GP', 'school_MS', 'sex_F', 'sex_M', 'age', 'address_R', 'address_U', 'famsize_GT3', 'famsize_LE3', 'Pstatus_A', 'Pstatus_T', 'Medu', 'Fedu', 'Mjob_at_home', 'Mjob_health', 'Mjob_other', 'Mjob_services', 'Mjob_teacher', 'Fjob_at_home', 'Fjob_health', 'Fjob_other', 'Fjob_services', 'Fjob_teacher', 'reason_course', 'reason_home', 'reason_other', 'reason_reputation', 'guardian_father', 'guardian_mother', 'guardian_other', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']


### Implementation: Training and Testing Data Split
So far, we have converted all _categorical_ features into numeric values. For the next step, we split the data (both features and corresponding labels) into training and test sets. In the following code cell below, you will need to implement the following:
- Randomly shuffle and split the data (`X_all`, `y_all`) into training and testing subsets.
  - Use 300 training points (approximately 75%) and 95 testing points (approximately 25%).
  - Set a `random_state` for the function(s) you use, if provided.
  - Store the results in `X_train`, `X_test`, `y_train`, and `y_test`.

In [5]:
# TODO: Import any additional functionality you may need here
from sklearn.model_selection import train_test_split
# TODO: Set the number of training points
num_train = 300

# Set the number of testing points
num_test = X_all.shape[0] - num_train

# TODO: Shuffle and split the dataset into the number of training and testing points above
X_train, X_test, y_train, y_test = train_test_split(X_all, y_all, test_size=num_test, train_size=num_train, random_state=0)

# Show the results of the split
print "Training set has {} samples.".format(X_train.shape[0])
print "Testing set has {} samples.".format(X_test.shape[0])

Training set has 300 samples.
Testing set has 95 samples.


## Training and Evaluating Models
In this section, you will choose 3 supervised learning models that are appropriate for this problem and available in `scikit-learn`. You will first discuss the reasoning behind choosing these three models by considering what you know about the data and each model's strengths and weaknesses. You will then fit the model to varying sizes of training data (100 data points, 200 data points, and 300 data points) and measure the F<sub>1</sub> score. You will need to produce three tables (one for each model) that shows the training set size, training time, prediction time, F<sub>1</sub> score on the training set, and F<sub>1</sub> score on the testing set.

**The following supervised learning models are currently available in** [`scikit-learn`](http://scikit-learn.org/stable/supervised_learning.html) **that you may choose from:**
- Gaussian Naive Bayes (GaussianNB)
- Decision Trees
- Ensemble Methods (Bagging, AdaBoost, Random Forest, Gradient Boosting)
- K-Nearest Neighbors (KNeighbors)
- Stochastic Gradient Descent (SGDC)
- Support Vector Machines (SVM)
- Logistic Regression

### Question 2 - Model Application
*List three supervised learning models that are appropriate for this problem. For each model chosen*
- Describe one real-world application in industry where the model can be applied. *(You may need to do a small bit of research for this — give references!)* 
- What are the strengths of the model; when does it perform well? 
- What are the weaknesses of the model; when does it perform poorly?
- What makes this model a good candidate for the problem, given what you know about the data?

**Answer: **<br>
The following three learning models are all use used for binary classification, hence, they are appropriate for this problem.

**1) KNN **
- This model is widely applied in Recommender Systems. Netflix is leveraging it to recommend new content to its customers. (https://www.quora.com/What-are-industry-applications-of-the-K-nearest-neighbor-algorithm) As long as people have watched a few movies that are the same and rated them (implicitly or explicitly), Netflix is able to leverage that feedback and suggest new movies based on other people's ratings and similarities between users' preferences.
- Strengths: 
    - robust to noisy data (it's instance based, only cares about the neighbours) 
    - simple to implement and easy to explain
- Weaknesses: 
    - classifying a new example/executing queries is quite expensive, depending on the number of neighbours and the complexity of the similarity function (since no training is happening upfront)
    - KNN is very affceted by the curse of dimensionality: exponentially more data is needed for every extra feature introduced
    - reseracher needs to make a lot of decisions that can influence the output of the algorithm (choice of similarity/distance metric, the number of neighbors to consider, voting rules and breaking ties)
- KNN is a good candidate for solving the problem at hand because it is a classification algorithm and finding students **similar** to those who failed in the past seems like a reasonable thing to do (and KNN is based on the notion similarity). 

**2) Naive Bayes **
- Naive Bayes can be applied to spam filtering, so perhaps Google is using it to filter spam from gmail accounts. Naive Bayes is able to predict how probable a certain combination of words is in a real email, and if it is below a certian arbitrary threshold, the email is considered to be spam.
- Strengths:
    - elegant, easy to implement, efficient, empirically successful
    - inference is cheap (growing leanearly with the number of attributed --> handles a large number of features well)
    - the model has few parameters that can be easily estimated with labeled data (counting and dividing by total count)
    - handles missing attributes well
- Weaknesses:
    - in natural language processing, doesn't deal well with phrases that consist of multiple words that take a different meaning than each word in the phrase (like hockey team "Ottawa Senators")
    - need to pay attention to new values of attributes that have not been yet seen in the dataset, because estimated with the data probability of that value will be 0, hence the product will be 0 as well --> need to apply smoothing of probabilities
- Naive Bayes is a good candidate for solving the problem at hand because it is a classification algorithm and looks like we have 45 features after all the transformations - quite a large number for 300 training examples.

**3) SVM **
- SVM is used in bioinformatics for cancer classification ("Gene Selection for Cancer Classification using Support Vector Machines" http://link.springer.com/article/10.1023/A:1012487302797). Bioinformatic datasets are very complex and SVM is able to create very intricate decision boundaries using the kernel trick.
- Strengths:
    - performs well when the data has a clear margin between classes
    - able to create very complicated decision boundaries
- Weaknesses:
    - doesn't deal very well with large number of features and large datasets
    - is prone to over-fitting to the noise in large datasets
- SVM is a good candidate for solving the problem at hand because it is a classification algorithm, the dataset we are dealing with is not very large, I suspect that there may be a clear margin between the two classes I need to predict

### Setup
Run the code cell below to initialize three helper functions which you can use for training and testing the three supervised learning models you've chosen above. The functions are as follows:
- `train_classifier` - takes as input a classifier and training data and fits the classifier to the data.
- `predict_labels` - takes as input a fit classifier, features, and a target labeling and makes predictions using the F<sub>1</sub> score.
- `train_predict` - takes as input a classifier, and the training and testing data, and performs `train_clasifier` and `predict_labels`.
 - This function will report the F<sub>1</sub> score for both the training and testing data separately.

In [6]:
def train_classifier(clf, X_train, y_train):
    ''' Fits a classifier to the training data. '''
    
    # Start the clock, train the classifier, then stop the clock
    start = time()
    clf.fit(X_train, y_train)
    end = time()
    
    # Print the results
    print "Trained model in {:.4f} seconds".format(end - start)

    
def predict_labels(clf, features, target):
    ''' Makes predictions using a fit classifier based on F1 score. '''
    
    # Start the clock, make predictions, then stop the clock
    start = time()
    y_pred = clf.predict(features)
    end = time()
    
    # Print and return results
    print "Made predictions in {:.4f} seconds.".format(end - start)
    return f1_score(target.values, y_pred, pos_label='yes')


def train_predict(clf, X_train, y_train, X_test, y_test):
    ''' Train and predict using a classifer based on F1 score. '''
    
    # Indicate the classifier and the training set size
    print "Training a {} using a training set size of {}. . .".format(clf.__class__.__name__, len(X_train))
    
    # Train the classifier
    train_classifier(clf, X_train, y_train)
    
    # Print the results of prediction for both training and testing
    print "F1 score for training set: {:.4f}.".format(predict_labels(clf, X_train, y_train))
    print "F1 score for test set: {:.4f}.".format(predict_labels(clf, X_test, y_test))

### Implementation: Model Performance Metrics
With the predefined functions above, you will now import the three supervised learning models of your choice and run the `train_predict` function for each one. Remember that you will need to train and predict on each classifier for three different training set sizes: 100, 200, and 300. Hence, you should expect to have 9 different outputs below — 3 for each model using the varying training set sizes. In the following code cell, you will need to implement the following:
- Import the three supervised learning models you've discussed in the previous section.
- Initialize the three models and store them in `clf_A`, `clf_B`, and `clf_C`.
 - Use a `random_state` for each model you use, if provided.
 - **Note:** Use the default settings for each model — you will tune one specific model in a later section.
- Create the different training set sizes to be used to train each model.
 - *Do not reshuffle and resplit the data! The new training points should be drawn from `X_train` and `y_train`.*
- Fit each model with each training set size and make predictions on the test set (9 in total).  
**Note:** Three tables are provided after the following code cell which can be used to store your results.

In [7]:
# TODO: Import the three supervised learning models from sklearn
# from sklearn import model_A (KNN)
from sklearn.neighbors import KNeighborsClassifier
# from sklearn import model_B (Gaussian Naive Bayes)
from sklearn.naive_bayes import GaussianNB
# from sklearn import model_C (SVM)
from sklearn.svm import SVC

# TODO: Initialize the three models
clf_A = KNeighborsClassifier()
clf_B = GaussianNB()
clf_C = SVC(random_state=0)

# TODO: Set up the training set sizes
X_train_100 = X_train[:100]
y_train_100 = y_train[:100]

X_train_200 = X_train[:200]
y_train_200 = y_train[:200]

X_train_300 = X_train[:300]
y_train_300 = y_train[:300]

# TODO: Execute the 'train_predict' function for each classifier and each training set size
# train_predict(clf, X_train, y_train, X_test, y_test)
train_predict(clf_A, X_train_100, y_train_100, X_test, y_test)
print ""
train_predict(clf_A, X_train_200, y_train_200, X_test, y_test)
print ""
train_predict(clf_A, X_train_300, y_train_300, X_test, y_test)
print ""

train_predict(clf_B, X_train_100, y_train_100, X_test, y_test)
print ""
train_predict(clf_B, X_train_200, y_train_200, X_test, y_test)
print ""
train_predict(clf_B, X_train_300, y_train_300, X_test, y_test)
print ""

train_predict(clf_C, X_train_100, y_train_100, X_test, y_test)
print ""
train_predict(clf_C, X_train_200, y_train_200, X_test, y_test)
print ""
train_predict(clf_C, X_train_300, y_train_300, X_test, y_test)

Training a KNeighborsClassifier using a training set size of 100. . .
Trained model in 0.0350 seconds
Made predictions in 0.0000 seconds.
F1 score for training set: 0.7972.
Made predictions in 0.0160 seconds.
F1 score for test set: 0.7068.

Training a KNeighborsClassifier using a training set size of 200. . .
Trained model in 0.0000 seconds
Made predictions in 0.0000 seconds.
F1 score for training set: 0.8571.
Made predictions in 0.0000 seconds.
F1 score for test set: 0.7121.

Training a KNeighborsClassifier using a training set size of 300. . .
Trained model in 0.0000 seconds
Made predictions in 0.0000 seconds.
F1 score for training set: 0.8722.
Made predictions in 0.0160 seconds.
F1 score for test set: 0.7482.

Training a GaussianNB using a training set size of 100. . .
Trained model in 0.0000 seconds
Made predictions in 0.0000 seconds.
F1 score for training set: 0.8550.
Made predictions in 0.0000 seconds.
F1 score for test set: 0.7481.

Training a GaussianNB using a training set siz

### Tabular Results
Edit the cell below to see how a table can be designed in [Markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#tables). You can record your results from above in the tables provided.

** Classifer 1 - K Nearest Neighbors**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |0.0350|0.0160|0.7972|0.7068|
| 200               |0.0000|0.0000|0.8571|0.7121|
| 300               |0.0000|0.0160|0.8722|0.7482|

** Classifer 2 - Gaussian Naive Bayes**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |0.0000|0.0000|0.8550|0.7481|
| 200               |0.0000|0.0000|0.8321|0.7132|
| 300               |0.0000|0.0000|0.8088|0.7500|

** Classifer 3 - C-Support Vector Classification**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |0.0260|0.0000|0.8591|0.7838|
| 200               |0.0160|0.0160|0.8693|0.7755|
| 300               |0.0000|0.0000|0.8692|0.7586|

## Choosing the Best Model
In this final section, you will choose from the three supervised learning models the *best* model to use on the student data. You will then perform a grid search optimization for the model over the entire training set (`X_train` and `y_train`) by tuning at least one parameter to improve upon the untuned model's F<sub>1</sub> score. 

### Question 3 - Choosing the Best Model
*Based on the experiments you performed earlier, in one to two paragraphs, explain to the board of supervisors what single model you chose as the best model. Which model is generally the most appropriate based on the available data, limited resources, cost, and performance?*

**Answer: ** <br>
- As I explained in Question 2, I believe all 3 models are appropriate based on the nature of the problem (binary classification) and data available. 
- Moreover, all 3 models take very similar amount of time for training and prediction (all render results under one twentieth of a second), so the cost of training them and getting the preditions are comparable, they would use approximately the same amount of resourses.
- Their performance, measured by F1 score on the test data (the higher, the better), is relatively close too, but SVM seems to be doing better (having higher F1 score) than other models for each training set size. It does have a troubling trend of producing worse F1 scores on test data with the increase of the training set size. I think it could be due to overfitting, which I can try to fix by adjusting hyper-parameters.
- Hence, my final choice is an SVM model, because among otherwise very comparable models it produces the best F1 score.

### Question 4 - Model in Layman's Terms
*In one to two paragraphs, explain to the board of directors in layman's terms how the final model chosen is supposed to work. Be sure that you are describing the major qualities of the model, such as how the model is trained and how the model makes a prediction. Avoid using advanced mathematical or technical jargon, such as describing equations or discussing the algorithm implementation.*

**Answer: **<br>
Support Vector Machines is a binary classification model that will allow us to predict if a given student will need intervention in order to graduate or not. 

#### Training the model
---
SVM is **looking for a boundary** that will best separate the two types of data points (students, in our case). Best boundary is the one that lies equally far from both groups. In order to find the best boubdary, SVM first needs to find general area where a boundary can be drown and than it can draw the boundary right in the middle of it. That area is referred to a the **"margin"** and it is the space between the two classes. 
* Consider the image below. SVM will first maximize the distance (margin) between the two classes (represented by the squares and circles) and will then draw a line right in the middle. This line is as far from the squares as it is from the circles. Hence, it is the best boundary in this case. When new data comes in, if the data point ends up to the right of the line, it will classified as a circle. If it will be to the left of the line, it will be classified as the square.
![alt text](http://docs.opencv.org/2.4/_images/optimal-hyperplane.png)
* The data shown in the image above is called **linearly separable**, because a straight line can separate the classes. Of course, this is not always/rarely the case. Consider the example below, where no straight line can separate the two distinct classes. It seems that a circle will be able to separate the classes, and SVM uses the **"kernel trick"** in order to achive this.
  - Kernel is a function that takes our available linearly inseparable data as input and tranforms it by adding some extra dimention to the data, so that it becomes linearly separable. In the picture below, extra dimention is the the length to the center of the circle, so all triangles have a smaller length measure than all the circles. Hence, triangles will all be below a certain length and circles will all be above it. This means that somewhere around there a linear boundary can be drawn.
![](http://www.eric-kim.net/eric-kim-net/posts/1/imgs/data_2d_to_3d_hyperplane.png)

Due to the kernel trick, SVM can draw very intricate decision boundaries. After the boundaries are in place, the model is trained and ready to make predictions. 

#### Making predictions
---
Predictions are made based on where the new data point lands compared to the decision boundary (eg., left/right, within/outside the circle). The new point will be classified as the rest of the points to the same side of the coundary already are.

### Implementation: Model Tuning
Fine tune the chosen model. Use grid search (`GridSearchCV`) with at least one important parameter tuned with at least 3 different values. You will need to use the entire training set for this. In the code cell below, you will need to implement the following:
- Import [`sklearn.grid_search.GridSearchCV`](http://scikit-learn.org/0.17/modules/generated/sklearn.grid_search.GridSearchCV.html) and [`sklearn.metrics.make_scorer`](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html).
- Create a dictionary of parameters you wish to tune for the chosen model.
 - Example: `parameters = {'parameter' : [list of values]}`.
- Initialize the classifier you've chosen and store it in `clf`.
- Create the F<sub>1</sub> scoring function using `make_scorer` and store it in `f1_scorer`.
 - Set the `pos_label` parameter to the correct value!
- Perform grid search on the classifier `clf` using `f1_scorer` as the scoring method, and store it in `grid_obj`.
- Fit the grid search object to the training data (`X_train`, `y_train`), and store it in `grid_obj`.

In [22]:
# TODO: Import 'GridSearchCV' and 'make_scorer'
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import make_scorer

# TODO: Create the parameters list you wish to tune
parameters =  {'kernel':['linear', 'poly', 'rbf', 'sigmoid'],
              'C': [0.1, 0.5, 0.7, 1, 1.5, 2, 10],
              'gamma': ['auto', 0.1, 0.5, 0.7, 1, 1.5, 2, 10]}

# TODO: Initialize the classifier
clf = SVC()

# TODO: Make an f1 scoring function using 'make_scorer' 
f1_scorer = make_scorer(f1_score, pos_label='yes')

# TODO: Perform grid search on the classifier using the f1_scorer as the scoring method
grid_obj = GridSearchCV(clf, parameters, scoring = f1_scorer)

# TODO: Fit the grid search object to the training data and find the optimal parameters
grid_obj.fit(X_train, y_train)

# Get the estimator
clf = grid_obj.best_estimator_

# Report the final F1 score for training and testing after parameter tuning
print "Tuned model has a training F1 score of {:.4f}.".format(predict_labels(clf, X_train, y_train))
print "Tuned model has a testing F1 score of {:.4f}.".format(predict_labels(clf, X_test, y_test))

Made predictions in 0.0080 seconds.
Tuned model has a training F1 score of 0.9717.
Made predictions in 0.0020 seconds.
Tuned model has a testing F1 score of 0.7919.


### Question 5 - Final F<sub>1</sub> Score
*What is the final model's F<sub>1</sub> score for training and testing? How does that score compare to the untuned model?*
<br><br>
***Answer:***<br>
Final model's F<sub>1</sub> score is:
- 97.17% for training *(vs 86.92% for untuned)*
- 79.19% for testing *(vs 75.86% for untuned)* 
<br>

Both scores are higher for the final model than for the original untuned model. It appears that there may be some overfitting going on since the gap between training and testing scores is quite large (~18 pct points) and training score approaches 100%. However, testing score is also quite high and is higher than that of other models concidered. So this seems like a reasonable model to choose in the end.

**Answer: **

> **Note**: Once you have completed all of the code implementations and successfully answered each question above, you may finalize your work by exporting the iPython Notebook as an HTML document. You can do this by using the menu above and navigating to  
**File -> Download as -> HTML (.html)**. Include the finished document along with this notebook as your submission.